corehq/apps/smsbillables/models.py

```diff
@@ -366,22 +366,19 @@ def _get_gateway_fee(cls, backend_api_id, backend_id,
                     direct_gateway_fee, multipart_count = \
                         cls.get_charge_details_through_api(backend_instance, backend_message_id)
 
-                    gateway_fee = SmsGatewayFee.get_by_criteria(
-                        backend_api_id,
-                        direction,
-                    )
                 else:
                     log_smsbillables_error(
-                        "Could not create gateway fee for message %s: no backend_message_id" % couch_id
+                        "Could not create direct gateway fee for message %s: no backend_message_id" % couch_id
                     )
-            else:
-                gateway_fee = SmsGatewayFee.get_by_criteria(
-                    backend_api_id,
-                    direction,
-                    backend_instance=backend_id,
-                    country_code=country_code,
-                    national_number=national_number,
-                )
+
+            # always grab the gateway fee even if using an api
+            gateway_fee = SmsGatewayFee.get_by_criteria(
```

Comment:
Looks like a fee may be charged where it previously was not (approximately `if not is_gateway_billable or not backend_message_id`, although it's a little more complicated than that). It seems of potential concern that a fee would be charged if `is_gateway_billable` is false-y. From your explanation on the PR it sounds like you have thought through this, but if anything I'm raising here is surprising then maybe it's worth another consideration?

Reply:
I think you misread this as all of this logic is nested within the `if is_gateway_billable:` block. 

You are right that behavior changed in that a `gateway_fee` will attempt to be found even when a `backend_message_id` does not exist. I'm honestly not sure what the desired behavior is for this path. We will most likely find a `gateway_fee` that does not specify an amount and charge $0.00 for the message. If a `gateway_fee` exists with a valid amount, we will charge that amount, and would have charged that amount regardless of a valid `backend_message_id` since [the default is to use a `gateway_fee`](https://github.com/dimagi/commcare-hq/blob/7eac9ce76c14b38034bb0bb24231aa99e527222c/corehq/apps/smsbillables/models.py#L293-L302).

This could present an issue if we want to charge to the API, but find a `gateway_fee` that fits the message AND has a valid amount, in which case we default to the `gateway_fee`. However, my understanding is _this is the current behavior_, since we attempt to find a `gateway_fee` with very little criteria (no country_code) at the same time we set the `direct_gateway_fee`. 

Part of our process for managing backends/fees is to set a `gateway_fee` with a null amount when we want to use the API, so this would be an error on the developer's part. Whether or not that is a good process for managing prices seems like another discussion.  

That was a mouthful so let me know if I need to clarify anything.

Reply:
> I think you misread this as all of this logic is nested within the `if is_gateway_billable:` block.

You are correct, sorry for the confusion.

I'm assuming we do not have tests that cover these different branches and their requisite fee outcomes?

Reply:
@gherceg i think it might be good to add these tests. Also, as a side note, I started keeping track of instances where we attempt to pull the cost of the message from a gateway but it returns 0, thinking it would go away after the celery issues were unblocked https://sentry.io/organizations/dimagi/issues/1929862358/?project=136860&query=is%3Aunresolved this is of course beyond the scope of the ask of this ticket, but might be resolvable if the gateway (which I'm assuming is the gateway erroring out here) has a `gateway_fee` set.

Reply:
actually, after digging further, your work here would likely help resolve that error. would you be able to add a check here to see that if no charge is obtained and that if no gateway fee is set that the exception is thrown? because based on the work here this feels like the norm for Tropo gateways? https://github.com/dimagi/commcare-hq/blob/f24a68da093d42fa2fe34a4261617ba6c50000a0/corehq/apps/smsbillables/models.py#L408-L409

Reply:
The tests I added here aim to check the different paths to retrieving a price via `gateway_charge`. I can improve the naming and realizing there are a few missing cases after this conversation. I'll add those.

@biyeun and I offlined and concluded that the issue with an API price being $0 is out of the scope of this PR.